test/functional/p2p_ibd_stalling.py

```diff
@@ -0,0 +1,145 @@
+#!/usr/bin/env python3
+# Copyright (c) 2022- The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+"""
+Test stalling logic during IBD
+"""
+
+import time
+
+from test_framework.blocktools import (
+        create_block,
+        create_coinbase
+)
+from test_framework.messages import (
+        MSG_BLOCK,
+        MSG_TYPE_MASK,
+)
+from test_framework.p2p import (
+        CBlockHeader,
+        msg_block,
+        msg_headers,
+        P2PDataStore,
+)
+from test_framework.test_framework import BitcoinTestFramework
+from test_framework.util import (
+        assert_equal,
+)
+
+
+class P2PStaller(P2PDataStore):
+    def __init__(self, stall_block):
+        self.is_stalling = True
+        self.stall_block = stall_block
+        self.blocks_sent = 0
+        super().__init__()
+
+    def on_getdata(self, message):
+        for inv in message.inv:
+            self.getdata_requests.append(inv.hash)
+            if (inv.type & MSG_TYPE_MASK) == MSG_BLOCK:
+                if(not self.is_stalling or inv.hash != self.stall_block):
+                    self.send_message(msg_block(self.block_store[inv.hash]))
+                    self.blocks_sent += 1
+
+    def on_getheaders(self, message):
+        pass
+
+
+class P2PIBDStallingTest(BitcoinTestFramework):
+    def set_test_params(self):
+        self.setup_clean_chain = True
+        self.num_nodes = 1
+
+    def run_test(self):
+        NUM_BLOCKS = 1025
+        node = self.nodes[0]
+        tip = int(node.getbestblockhash(), 16)
+        blocks = []
+        height = 1
+        block_time = node.getblock(node.getbestblockhash())['time'] + 1
+        self.log.info("Prepare blocks without sending to the node")
+        block_dict = {}
+        for _ in range(NUM_BLOCKS):
+            blocks.append(create_block(tip, create_coinbase(height), block_time))
+            blocks[-1].solve()
+            tip = blocks[-1].sha256
+            block_time += 1
+            height += 1
+            block_dict[blocks[-1].sha256] = blocks[-1]
+
+        stall_block = blocks[0].sha256
+
+        headers_message = msg_headers()
+        headers_message.headers = [CBlockHeader(b) for b in blocks[:NUM_BLOCKS-1]]
+        peers = []
+
+        self.log.info("Check that stalling does not start the stalling logic yet if there is a 1024 block window.")
+        with self.nodes[0].assert_debug_log([], unexpected_msgs=['Stall started']):
+            for id in range(8):
+                peers.append(node.add_outbound_p2p_connection(P2PStaller(stall_block), p2p_idx=id, connection_type="outbound-full-relay"))
+                peers[-1].block_store = block_dict
+                peers[-1].send_message(headers_message)
+            self.wait_until(lambda: self.total_blocks_sent(peers) == NUM_BLOCKS - 2)
+        time.sleep(0.5)  # Wait until all blocks have arrived at the node
```

Comment:
Not a fan of timeouts like this but in case i also don't see how to avoid it.

Reply:
yes, I was looking for an RPC we could `wait_until` for to avoid this - but I didn't find a way of querying the number of blocks a node has downloaded (including those not connected to the chain yet).

Reply:
`getpeerinfo()[0]["bytesrecv_per_msg"]["block"] ?`

Reply:
To follow @glozow's idea, this should work:

```diff
diff --git i/test/functional/p2p_ibd_stalling.py w/test/functional/p2p_ibd_stalling.py
index 78626c003b..d593187d74 100755
--- i/test/functional/p2p_ibd_stalling.py
+++ w/test/functional/p2p_ibd_stalling.py
@@ -79,13 +79,14 @@ class P2PIBDStallingTest(BitcoinTestFramework):
         with self.nodes[0].assert_debug_log([], unexpected_msgs=['Stall started']):
             for id in range(8):
                 peers.append(node.add_outbound_p2p_connection(P2PStaller(stall_block), p2p_idx=id, connection_type="outbound-full-relay"))
                 peers[-1].block_store = block_dict
                 peers[-1].send_message(headers_message)
             self.wait_until(lambda: self.total_blocks_sent(peers) == NUM_BLOCKS - 2)
-        time.sleep(0.5)  # Wait until all blocks have arrived at the node
+            self.wait_until(lambda: self.total_bytes_recv_for_blocks() == 172761)
+
 
         self.log.info("Check that increasing the window beyond 1024 blocks triggers stalling logic")
         headers_message.headers = [CBlockHeader(b) for b in blocks]
         with self.nodes[0].assert_debug_log(expected_msgs=['Stall started peer=0']):
             for p in peers:
                 p.send_message(headers_message)
@@ -149,9 +150,15 @@ class P2PIBDStallingTest(BitcoinTestFramework):
     def total_blocks_sent(self, peers):
         num_blocks = 0
         for p in peers:
             num_blocks += p.blocks_sent
         return num_blocks
 
+    def total_bytes_recv_for_blocks(self):
+        total = 0
+        for info in self.nodes[0].getpeerinfo():
+            total += info["bytesrecv_per_msg"]["block"]
+        return total
+
 
 if __name__ == '__main__':
     P2PIBDStallingTest().main()
```

and even better if we can put some formula behind the magic number.

Reply:
or, more pythonish:

```python
self.wait_until(lambda: sum(e["bytesrecv_per_msg"]["block"] for e in self.nodes[0].getpeerinfo()) == 172761)
```

Reply:
Thanks, that works!
We send 1023 blocks, the formula would be `126 * 168 + 897 * 169 = 172761` (at some point the blocks get larger by 1 byte).
Not super stable because `bytesrecv_per_msg` includes the extra 24 bytes magic etc. (not just the payload) - so I think the magic number would e.g. be different for BIP324 - but it's definitely better than a fixed timeout.

Reply:
I ended up using the less pythonish version because we also need a check that the "block" field exists in "bytesrecv_per_msg", and I find that easier to read. Added a comment for the magic number.